File: 2021/day10/day10.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stack>
#include <unordered_map>

using namespace std;

typedef long long ll;

void part1(vector<string> lines);
void part2(vector<string> lines);
vector<string> readInput();
pair<unordered_map<char, int>, vector<stack<char>>> getAndFilterCorrupted(vector<string> &lines);
int getCorruptedScore(unordered_map<char, int> &corruptedCount);
ll getAutocompleteScore(stack<char> &incompleteStack);
vector<ll> getAutocompleteScores(vector<stack<char>> &incompleteStacks);
ll getMiddleValue(vector<ll> &scores);


int main() {
    vector<string> lines = readInput();
    part1(lines);
    part2(lines);
    return 0;
}


void part1(vector<string> lines) {
    auto output = getAndFilterCorrupted(lines);
    int score = getCorruptedScore(output.first);
    cout << "Part 1: " << score << endl;
}

void part2(vector<string> lines) {
    auto output = getAndFilterCorrupted(lines);
    auto scores = getAutocompleteScores(output.second);
    ll middleScore = getMiddleValue(scores);
    cout << "Part 2: " << middleScore << endl;
}

vector<string> readInput() {
    vector<string> lines;
    string line;
    while (cin >> line) {
        lines.push_back(line);
    }
    return lines;
}

pair<unordered_map<char, int>, vector<stack<char>>> getAndFilterCorrupted(vector<string> &lines) {
    unordered_map<char, int> corruptedCount = {{')', 0}, {']', 0}, {'}', 0}, {'>', 0}};
    vector<stack<char>> incompleteStacks;
    for (const string& line : lines) {
        stack<char> bracketStack;
        bool isCorrupted = false;
        for (char bracket : line) {
            string openBrackets = "([{<";
            if (openBrackets.find(bracket) != string::npos) {
                bracketStack.push(bracket);
            } else if ((bracket == ')' && bracketStack.top() == '(') ||
                       (bracket == ']' && bracketStack.top() == '[') ||
                       (bracket == '}' && bracketStack.top() == '{') ||
                       (bracket == '>' && bracketStack.top() == '<')) {
                bracketStack.pop();
            } else {
                corruptedCount[bracket]++;
                isCorrupted = true;
                break;
            }
        }
        if (!isCorrupted) {
            incompleteStacks.push_back(bracketStack);
        }
    }
    pair<unordered_map<char, int>, vector<stack<char>>> output;
    output.first = corruptedCount;
    output.second = incompleteStacks;
    return output;
}

int getCorruptedScore(unordered_map<char, int> &corruptedCount) {
    int score = corruptedCount[')'] * 3 +
                corruptedCount[']'] * 57 +
                corruptedCount['}'] * 1197 +
                corruptedCount['>'] * 25137;
    return score;
}

ll getAutocompleteScore(stack<char> &incompleteStack) {
    unordered_map<char, int> points = {{'(', 1}, {'[', 2}, {'{', 3}, {'<', 4}};
    ll score = 0;
    while (!incompleteStack.empty()) {
        char bracket = incompleteStack.top();
        incompleteStack.pop();
        score *= 5;
        score += points[bracket];
    }
    return score;
}

vector<ll> getAutocompleteScores(vector<stack<char>> &incompleteStacks) {
    vector<ll> scores;
    for (auto stack : incompleteStacks) {
        ll points = getAutocompleteScore(stack);
        scores.push_back(points);
    }
    return scores;
}

ll getMiddleValue(vector<ll> &scores) {
    sort(scores.begin(), scores.end());
    return scores[scores.size() / 2];
}

File: 2021/day10/day10.py
def read_input():
    file = open('input.in', 'r')
    lines = [line.strip() for line in file.readlines()]
    return lines


def get_and_filter_corrupted(lines):
    corrupted_count = {')': 0, ']': 0, '}': 0, '>': 0}
    corrupted_index = []
    for i in range(len(lines)):
        line = lines[i]
        stack = []
        for bracket in line:
            if bracket == '(' or bracket == '[' or bracket == '{' or bracket == '<':
                stack.append(bracket)
            elif (bracket == ')' and stack[-1] == '(') or\
                    (bracket == ']' and stack[-1] == '[') or\
                    (bracket == '}' and stack[-1] == '{') or\
                    (bracket == '>' and stack[-1] == '<'):
                del stack[-1]
            else:
                corrupted_count[bracket] += 1
                corrupted_index.append(i)
                break
    for i in corrupted_index[::-1]:
        del lines[i]
    return corrupted_count


def get_corrupted_score(corrupted_count):
    score = corrupted_count[')'] * 3 + \
            corrupted_count[']'] * 57 + \
            corrupted_count['}'] * 1197 + \
            corrupted_count['>'] * 25137
    return score


def auto_complete(lines):
    completed_lines = []
    for line in lines:
        stack = []
        for bracket in line:
            if bracket == '(' or bracket == '[' or bracket == '{' or bracket == '<':
                stack.append(bracket)
            elif (bracket == ')' and stack[-1] == '(') or \
                    (bracket == ']' and stack[-1] == '[') or \
                    (bracket == '}' and stack[-1] == '{') or \
                    (bracket == '>' and stack[-1] == '<'):
                del stack[-1]
            else:
                print("ERROR")
                break
        completed_lines.append(stack)
    return completed_lines


def get_score(stack):
    points = {'(': 1, '[': 2, '{': 3, '<': 4}
    score = 0
    for bracket in stack[::-1]:
        score *= 5
        score += points[bracket]
    return score


def get_scores(stacks):
    return [get_score(stack) for stack in stacks]


def get_middle_score(scores):
    scores.sort()
    return scores[len(scores)//2]


def part1():
    lines = read_input()
    corrupted_count = get_and_filter_corrupted(lines)
    score = get_corrupted_score(corrupted_count)
    print(f"Part 1: {score}")


def part2():
    lines = read_input()
    _ = get_and_filter_corrupted(lines)
    stacks = auto_complete(lines)
    scores = get_scores(stacks)
    middle_score = get_middle_score(scores)
    print(f"Part 2: {middle_score}")


if __name__ == "__main__":
    part1()
    part2()
